### ISSUE TITLE
Build task running VS2017RC does not find UnitTestFramework

### ISSUE THREAD

**Issue by Pitterling** [Microsoft/vsts-tasks#3359]
local agent - VS2017RC installed - Build task does not find UnitTestFramework and therefore all *.test.dlls are not built. building the solution within VS2017RC on same machine succeeds.

> C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\amd64\Microsoft.Common.CurrentVersion.targets(1909,5): Warning MSB3245: Could not resolve this reference. Could not locate the assembly "Microsoft.VisualStudio.QualityTools.UnitTestFramework". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors. 

> 2016-12-21T23:57:38.7723245Z CoreCompile:
2016-12-21T23:57:38.7723245Z   C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\amd64\vbc.exe /noconfig /imports:System,System.Collections,System.Collections.Generic,System.Diagnostics,System.Linq,System.Xml.Linq,Microsoft.VisualBasic,Microsoft.VisualStudio.TestTools.UnitTesting /optioncompare:Binary /optionexplicit+ /optionstrict+ /optioninfer+ /nostdlib /platform:AnyCPU /rootnamespace:AIX_io_lunpath_details.Console.Tests /sdkpath:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5.2" /highentropyva+ /doc:obj\Debug\AIX_io_lunpath_details.Console.Tests.xml /define:"CONFIG=\"Debug\",DEBUG=-1,TRACE=-1,_MyType=\"Windows\",PLATFORM=\"AnyCPU\"" /reference:C:\TfsBuildAgents\_work\1\s\bin\Debug\AIX_io_lunpath_details.Console.exe,C:\TfsBuildAgents\_work\1\s\bin\Debug\AIX_io_lunpath_details.Service.dll,C:\TfsBuildAgents\_work\1\s\packages\CommandLineParser.2.1.1-beta\lib\net45\CommandLine.dll,C:\TfsBuildAgents\_work\1\s\packages\NLog.4.3.11\lib\net45\NLog.dll,"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5.2\System.Core.dll","C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5.2\System.dll","C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5.2\System.Xml.dll","C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5.2\System.Xml.Linq.dll" /debug+ /debug:full /filealign:512 /out:obj\Debug\AIX_io_lunpath_details.Console.Tests.dll /ruleset:..\TOP10+MS_Managed_Minimum.ruleset /subsystemversion:6.00 /target:library /utf8output CmdLineOptionsParserTests.vb "My Project\AssemblyInfo.vb" "C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Temp\.NETFramework,Version=v4.5.2.AssemblyAttributes.vb" /warnaserror+:41999,42016,42017,42018,42019,42020,42021,42022,42032,42036
2016-12-21T23:57:38.7723245Z   Using shared compilation with compiler from directory: C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin
2016-12-21T23:57:38.8504505Z ##[warning]vbc(0,0): Warning BC40057: Namespace or type specified in the project-level Imports 'Microsoft.VisualStudio.TestTools.UnitTesting' doesn't contain any public member or cannot be found. Make sure the namespace or the type is defined and contains at least one public member. Make sure the imported element name doesn't use any aliases.
2016-12-21T23:57:38.8504505Z ##[debug]Processed: ##vso[task.logissue type=Warning;sourcepath=vbc;linenumber=0;columnnumber=0;code=BC40057;]Namespace or type specified in the project-level Imports 'Microsoft.VisualStudio.TestTools.UnitTesting' doesn't contain any public member or cannot be found. Make sure the namespace or the type is defined and contains at least one public member. Make sure the imported element name doesn't use any aliases.
2016-12-21T23:57:38.8504505Z vbc : warning BC40057: Namespace or type specified in the project-level Imports 'Microsoft.VisualStudio.TestTools.UnitTesting' doesn't contain any public member or cannot be found. Make sure the namespace or the type is defined and contains at least one public member. Make sure the imported element name doesn't use any aliases. [C:\TfsBuildAgents\_work\1\s\AIX_io_lunpath_details.Console.Tests\AIX_io_lunpath_details.Console.Tests.vbproj]

**Comment by kaadhina** [comment]
Which task are you using: MSBuild/VSBuild ?
What option are you selecting for VS version /MSBuild version in the drop down?

**Comment by Pitterling** [comment]
@kaadhina Visual Studio Build
![image](https://cloud.githubusercontent.com/assets/4181164/21766629/ec40f4dc-d66f-11e6-8066-32452f556051.png)

**Comment by AbhitejJohn** [comment]
@Pitterling : Can we use msbuild x86 instead? That is what Visual Studio uses to build the project in Dev15. 
MSbuild x64 does not seem to be resolving assemblies from Public Assemblies. This is however working in Dev14 and the x86 msbuild from Dev15 where the UnitTest Framework dll is being picked up from PublicAssemblies. @Microsoft/msbuild-maintainers, any idea why this resolution is not happening with msbuild x64?

**Comment by rainersigwald** [comment]
Filed https://github.com/Microsoft/msbuild/issues/1549 to investigate why there's an x64/x86 difference.

@Pitterling I also recommend using the default x86 MSBuild as a workaround.

**Comment by Pitterling** [comment]
@AbhitejJohn @rainersigwald Thanks for the update. I changed to x86 but also updated the VS2015RC version. Now VS15 is not found any longer :( ... therefore I cannot confirm if this changes the issue.

**Comment by ericsciple** [comment]
@Pitterling are you on VSTS (the service) or an on-premises install of TFS?

**Comment by Pitterling** [comment]
@ericsciple TF2017 - agent/vs2017rc running on same node - recently updated VS2017RC to 15.0.26020.0. With former VS15 version agent detected my VS2017RC installation.

**Comment by ericsciple** [comment]
can you add the variable `system.debug` to your definition and set the value to `true`, run a build, and download the build logs zip and send to ersciple? (at microsoft com)

**Comment by ericsciple** [comment]
i can't repro the issue with build 26020. are you sure the installation succeeded? can you open visual studio on the box?

**Comment by Pitterling** [comment]
@ericsciple yes, installation succeeded, i can open vs on the box.

Agent.Version = 2.111.1 - Capabilities tab does not list "VisualStudio_15.0"

**Comment by ericsciple** [comment]
ok, the task finding VS and the agent finding VS are two different things. the agent scans on startup. so you may just need to restart the agent.

**Comment by Pitterling** [comment]
@ericsciple I restarted several times, I also reverted back the agent to 2.109.2 ... VS15 is not found.

**Comment by ericsciple** [comment]
When you launch the visual studio installer, does it know where your instance of VS 2017 is located? When you click modify on the instance, is the "Location" text box filled in corrected (bottom of workloads window)

**Comment by ericsciple** [comment]
The only other thing I can think of, if the agent is running as a service account, does the service account have access to the VS installation directory? For example, if you run the agent as Network Service, then it wouldn't be able to read files under your user profile directory.

**Comment by Pitterling** [comment]
@ericsciple yep, Installer knows the correct location and i tried both user/service - no luck.

Is there any way to trace the "scanning capabilities" phase during agent startup?

**Comment by ericsciple** [comment]
Let me pull together a small script to help diagnose. We may need someone from VS to help investigate.

**Comment by ericsciple** [comment]
Ok here's as small as I could reduce it. Log on the box with VS 2017 installed, add the below code into a script Test-VS2017.ps1, and then CD to that directory and run .\Test-VS2017.ps1

It sounds like it will output 0 instances for you. VS 2017 doesn't have registry values to locate where it's installed, so we query the COM API instead. You could also try running it elevated and see if that makes a difference.

```powershell
param([switch]$Nested)

if (!$Nested)
{
    & powershell.exe $PSCommandPath -Nested
    return
}

Import-Module -Name 'Microsoft.PowerShell.Management'
Import-Module -Name 'Microsoft.PowerShell.Utility'
$ErrorActionPreference = 'Stop'

if (!(Test-Path -LiteralPath 'REGISTRY::HKEY_CLASSES_ROOT\CLSID\{177F0C4A-1CD3-4DE7-A32C-71DBBB9FA36D}')) {
    Write-Host "Not found: HKCR\CLSID\{177F0C4A-1CD3-4DE7-A32C-71DBBB9FA36D}"
    return
}

Write-Host "Found: HKCR\CLSID\{177F0C4A-1CD3-4DE7-A32C-71DBBB9FA36D}"
Write-Host "Adding Visual Studio setup helpers."
Add-Type -Debug:$false -TypeDefinition @'
namespace CapabilityHelpers.VisualStudio.Setup
{
    using System;
    using System.Collections.Generic;
    using CapabilityHelpers.VisualStudio.Setup.Com;

    public sealed class Instance
    {
        public string Description { get; set; }

        public string DisplayName { get; set; }

        public string Id { get; set; }

        public System.Runtime.InteropServices.ComTypes.FILETIME InstallDate { get; set; }

        public string Name { get; set; }

        public string Path { get; set; }

        public Version Version
        {
            get
            {
                try
                {
                    return new Version(VersionString);
                }
                catch (Exception)
                {
                    return new Version(0, 0);
                }
            }
        }

        public string VersionString { get; set; }

        public static List<Instance> GetInstances()
        {
            List<Instance> list = new List<Instance>();
            ISetupConfiguration config = new SetupConfiguration() as ISetupConfiguration;
            IEnumSetupInstances enumInstances = config.EnumInstances();
            ISetupInstance[] instances = new ISetupInstance[1];
            int fetched = 0;
            enumInstances.Next(1, instances, out fetched);
            while (fetched > 0)
            {
                ISetupInstance instance = instances[0];
                list.Add(new Instance()
                {
                    Description = instance.GetDescription(),
                    DisplayName = instance.GetDisplayName(),
                    Id = instance.GetInstanceId(),
                    InstallDate = instance.GetInstallDate(),
                    Name = instance.GetInstallationName(),
                    Path = instance.GetInstallationPath(),
                    VersionString = instance.GetInstallationVersion(),
                });
                enumInstances.Next(1, instances, out fetched);
            }

            return list;
        }
    }
}

namespace CapabilityHelpers.VisualStudio.Setup.Com
{
    using System;
    using System.Runtime.InteropServices;

    [ComImport]
    [Guid("6380BCFF-41D3-4B2E-8B2E-BF8A6810C848")]
    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
    public interface IEnumSetupInstances
    {
        void Next(
            [In, MarshalAs(UnmanagedType.U4)] int celt,
            [Out, MarshalAs(UnmanagedType.LPArray, ArraySubType = UnmanagedType.Interface)] ISetupInstance[] rgelt,
            [Out, MarshalAs(UnmanagedType.U4)] out int pceltFetched);

        void Skip([In, MarshalAs(UnmanagedType.U4)] int celt);

        void Reset();

        [return: MarshalAs(UnmanagedType.Interface)]
        IEnumSetupInstances Clone();
    }

    [ComImport]
    [Guid("42843719-DB4C-46C2-8E7C-64F1816EFD5B")]
    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
    public interface ISetupConfiguration
    {
        [return: MarshalAs(UnmanagedType.Interface)]
        IEnumSetupInstances EnumInstances();
    }

    [ComImport]
    [Guid("B41463C3-8866-43B5-BC33-2B0676F7F42E")]
    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
    public interface ISetupInstance
    {
        [return: MarshalAs(UnmanagedType.BStr)]
        string GetInstanceId();

        [return: MarshalAs(UnmanagedType.Struct)]
        System.Runtime.InteropServices.ComTypes.FILETIME GetInstallDate();

        [return: MarshalAs(UnmanagedType.BStr)]
        string GetInstallationName();

        [return: MarshalAs(UnmanagedType.BStr)]
        string GetInstallationPath();

        [return: MarshalAs(UnmanagedType.BStr)]
        string GetInstallationVersion();

        [return: MarshalAs(UnmanagedType.BStr)]
        string GetDisplayName([In, MarshalAs(UnmanagedType.U4)] int lcid = default(int));

        [return: MarshalAs(UnmanagedType.BStr)]
        string GetDescription([In, MarshalAs(UnmanagedType.U4)] int lcid = default(int));
    }

    [ComImport]
    [Guid("42843719-DB4C-46C2-8E7C-64F1816EFD5B")]
    [CoClass(typeof(SetupConfigurationClass))]
    [TypeLibImportClass(typeof(SetupConfigurationClass))]
    public interface SetupConfiguration : ISetupConfiguration
    {
    }

    [ComImport]
    [Guid("177F0C4A-1CD3-4DE7-A32C-71DBBB9FA36D")]
    [ClassInterface(ClassInterfaceType.None)]
    public class SetupConfigurationClass
    {
    }
}
'@
Write-Host "Getting Visual Studio setup instances."
$instances = @( [CapabilityHelpers.VisualStudio.Setup.Instance]::GetInstances() )
Write-Host "Found $($instances.Count) instances."
Write-Host ($instances | Format-List * | Out-String)
```

**Comment by Pitterling** [comment]
here is the output

> PS C:\Users\Administrator\Downloads> C:\Users\Administrator\Downloads\Test-VS2017.ps1
> Found: HKCR\CLSID\{177F0C4A-1CD3-4DE7-A32C-71DBBB9FA36D}
> Adding Visual Studio setup helpers.
> Getting Visual Studio setup instances.
> Found 0 instances.

**Comment by ericsciple** [comment]
Thanks... I'll try and track down someone from VS who can help troubleshoot why the COM API isn't returning any instances, yet the VS installer is able to resolve the location.

**Comment by heaths** [comment]
Depending on the state of the instance, `ISetupConfiguration::EnumInstances` may not enumerate them (I only enumerate valid instances). You can use `ISetupConfiguration2::EnumInstances` instead to enumerate them all. Starting with RC.3 `ISetupConfiguration::GetInstanceForPath` will also return an instance for an installation directory (or subdirectory) even if the state is incomplete.

**Comment by ericsciple** [comment]
@Pitterling have you tried rebooting the machine?

**Comment by ericsciple** [comment]
@Pitterling also can you try this script, it will list more info about the state of the install

```powershell
param([switch]$Nested)

if (!$Nested)
{
    & powershell.exe $PSCommandPath -Nested
    return
}

Import-Module -Name 'Microsoft.PowerShell.Management'
Import-Module -Name 'Microsoft.PowerShell.Utility'
$ErrorActionPreference = 'Stop'

if (!(Test-Path -LiteralPath 'REGISTRY::HKEY_CLASSES_ROOT\CLSID\{177F0C4A-1CD3-4DE7-A32C-71DBBB9FA36D}')) {
    Write-Host "Not found: HKCR\CLSID\{177F0C4A-1CD3-4DE7-A32C-71DBBB9FA36D}"
    return
}

Write-Host "Found: HKCR\CLSID\{177F0C4A-1CD3-4DE7-A32C-71DBBB9FA36D}"
Write-Host "Adding Visual Studio setup helpers."
Add-Type -Debug:$false -TypeDefinition @'
namespace CapabilityHelpers.VisualStudio.Setup
{
    using System;
    using System.Collections.Generic;
    using CapabilityHelpers.VisualStudio.Setup.Com;

    public sealed class Instance
    {
        public string Description { get; set; }

        public string DisplayName { get; set; }

        public string Id { get; set; }

        public System.Runtime.InteropServices.ComTypes.FILETIME InstallDate { get; set; }

        public string Name { get; set; }

        public string Path { get; set; }

        public InstanceState State { get; set; }

        public Version Version
        {
            get
            {
                try
                {
                    return new Version(VersionString);
                }
                catch (Exception)
                {
                    return new Version(0, 0);
                }
            }
        }

        public string VersionString { get; set; }

        public static List<Instance> GetInstances()
        {
            List<Instance> list = new List<Instance>();
            ISetupConfiguration2 config = new SetupConfiguration() as ISetupConfiguration2;
            IEnumSetupInstances enumInstances = config.EnumAllInstances();
            ISetupInstance[] instances = new ISetupInstance[1];
            int fetched = 0;
            enumInstances.Next(1, instances, out fetched);
            while (fetched > 0)
            {
                ISetupInstance2 instance = instances[0] as ISetupInstance2;
                list.Add(new Instance()
                {
                    Description = instance.GetDescription(),
                    DisplayName = instance.GetDisplayName(),
                    Id = instance.GetInstanceId(),
                    InstallDate = instance.GetInstallDate(),
                    Name = instance.GetInstallationName(),
                    Path = instance.GetInstallationPath(),
                    State = instance.GetState(),
                    VersionString = instance.GetInstallationVersion(),
                });
                enumInstances.Next(1, instances, out fetched);
            }

            return list;
        }
    }
}

namespace CapabilityHelpers.VisualStudio.Setup.Com
{
    using System;
    using System.Runtime.InteropServices;

    [ComImport]
    [Guid("6380BCFF-41D3-4B2E-8B2E-BF8A6810C848")]
    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
    public interface IEnumSetupInstances
    {
        void Next(
            [In, MarshalAs(UnmanagedType.U4)] int celt,
            [Out, MarshalAs(UnmanagedType.LPArray, ArraySubType = UnmanagedType.Interface)] ISetupInstance[] rgelt,
            [Out, MarshalAs(UnmanagedType.U4)] out int pceltFetched);

        void Skip([In, MarshalAs(UnmanagedType.U4)] int celt);

        void Reset();

        [return: MarshalAs(UnmanagedType.Interface)]
        IEnumSetupInstances Clone();
    }

    [ComImport]
    [Guid("42843719-DB4C-46C2-8E7C-64F1816EFD5B")]
    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
    public interface ISetupConfiguration
    {
        [return: MarshalAs(UnmanagedType.Interface)]
        IEnumSetupInstances EnumInstances();
    }

    [ComImport]
    [Guid("B41463C3-8866-43B5-BC33-2B0676F7F42E")]
    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
    public interface ISetupInstance
    {
        [return: MarshalAs(UnmanagedType.BStr)]
        string GetInstanceId();

        [return: MarshalAs(UnmanagedType.Struct)]
        System.Runtime.InteropServices.ComTypes.FILETIME GetInstallDate();

        [return: MarshalAs(UnmanagedType.BStr)]
        string GetInstallationName();

        [return: MarshalAs(UnmanagedType.BStr)]
        string GetInstallationPath();

        [return: MarshalAs(UnmanagedType.BStr)]
        string GetInstallationVersion();

        [return: MarshalAs(UnmanagedType.BStr)]
        string GetDisplayName([In, MarshalAs(UnmanagedType.U4)] int lcid = default(int));

        [return: MarshalAs(UnmanagedType.BStr)]
        string GetDescription([In, MarshalAs(UnmanagedType.U4)] int lcid = default(int));
    }

    [ComImport]
    [Guid("42843719-DB4C-46C2-8E7C-64F1816EFD5B")]
    [CoClass(typeof(SetupConfigurationClass))]
    [TypeLibImportClass(typeof(SetupConfigurationClass))]
    public interface SetupConfiguration : ISetupConfiguration, ISetupConfiguration2
    {
    }

    [ComImport]
    [Guid("177F0C4A-1CD3-4DE7-A32C-71DBBB9FA36D")]
    [ClassInterface(ClassInterfaceType.None)]
    public class SetupConfigurationClass
    {
    }

    [Flags]
    public enum InstanceState : uint
    {
        None = 0u,
        Local = 1u,
        Registered = 2u,
        NoRebootRequired = 4u,
        Complete = 4294967295u
    }

    [ComImport]
    [Guid("89143C9A-05AF-49B0-B717-72E218A2185C")]
    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
    public interface ISetupInstance2 : ISetupInstance
    {
        [return: MarshalAs(UnmanagedType.BStr)]
        new string GetInstanceId();

        [return: MarshalAs(UnmanagedType.Struct)]
        new System.Runtime.InteropServices.ComTypes.FILETIME GetInstallDate();

        [return: MarshalAs(UnmanagedType.BStr)]
        new string GetInstallationName();

        [return: MarshalAs(UnmanagedType.BStr)]
        new string GetInstallationPath();

        [return: MarshalAs(UnmanagedType.BStr)]
        new string GetInstallationVersion();

        [return: MarshalAs(UnmanagedType.BStr)]
        new string GetDisplayName([MarshalAs(UnmanagedType.U4)] [In] int lcid = 0);

        [return: MarshalAs(UnmanagedType.BStr)]
        new string GetDescription([MarshalAs(UnmanagedType.U4)] [In] int lcid = 0);

        [return: MarshalAs(UnmanagedType.BStr)]
        string ResolvePath([MarshalAs(UnmanagedType.LPWStr)] [In] string pwszRelativePath = null);

        [return: MarshalAs(UnmanagedType.U4)]
        InstanceState GetState();
    }

    [ComImport]
    [Guid("26AAB78C-4A60-49D6-AF3B-3C35BC93365D")]
    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
    public interface ISetupConfiguration2
    {
        [return: MarshalAs(UnmanagedType.Interface)]
        IEnumSetupInstances EnumAllInstances();
    }
}
'@
Write-Host "Getting Visual Studio setup instances."
$instances = @( [CapabilityHelpers.VisualStudio.Setup.Instance]::GetInstances() )
Write-Host "Found $($instances.Count) instances."
Write-Host ($instances | Format-List * | Out-String)
```

**Comment by Pitterling** [comment]
@ericsciple i don't see any difference in the output of both scripts. I also rebooted the system.

> PS C:\Users\Administrator\Downloads> C:\Users\Administrator\Downloads\Test-VS2017-2.ps1
> Found: HKCR\CLSID\{177F0C4A-1CD3-4DE7-A32C-71DBBB9FA36D}
> Adding Visual Studio setup helpers.
> Getting Visual Studio setup instances.
>
> Found 0 instances.

**Comment by ericsciple** [comment]
@heaths ISetupConfiguration2.EnumAllInstances is also returning zero instances. Thoughts?

**Comment by heaths** [comment]
Please run http://aka.ms/vscollect.exe and share _%TEMP%\vslogs.zip_ after it completes.

**Comment by Pitterling** [comment]
@heaths file is about 20MB .. any preferred place to upload?

**Comment by heaths** [comment]
OneDrive, Dropbox, wherever. Just make sure I can access it. I can let you know once I've downloaded if you want to delete it afterward.

**Comment by Pitterling** [comment]
@heaths [vslogs](https://www.dropbox.com/sh/51c52looitssanl/AABU1s-Fe3Pbi5b9nnNJ9rJ5a?dl=0) contains 2 files
- TF-agent - the one of interest
- laptop --> my laptop, powershell script does not detect VS15 either

**Comment by heaths** [comment]
Thanks. I've got them and am taking a look now.

**Comment by heaths** [comment]
I don't see anything wrong other than some errors, which newer versions of the API (like that shipped with RC2) should still enumerate as some errors still allow the product to be launchable. Depending on which version of the DLLs the original code (from your OP) called, this could be the problem. But trying the script with your state yielded the same results you got (even though my official powershell cmdlets I hope to release soon work), which points to the problem that the script @ericsciple posted is not correct. The methods have to be in the correct VTBL order and are not. In fact, `ISetupConfiguration2` doesn't contain (and can't simply do it through interface inheritance) `ISetupConfiguration`'s methods, which it must or the script is basically calling the wrong method.

In the "laptop" files, some packages related to Azure failed and related packages were skipped. In the "TestAgent" files, web tools failed and related packages were skipped. Newer versions of the API should work correctly.

RC3 is coming out soon if you can wait; otherwise, you can open the Visual Studio Installer and try to repair the issues which should get the original code working.

As for the test script, if you can wait I should be releasing our official supported cmdlets soon; otherwise, I can fix the script in this post but it will only enumerate instances. It won't fix the problem.

**Comment by ericsciple** [comment]
Thanks @heaths! To add more context, the task in VSTS uses `ISetupConfiguration` to locate msbuild and detect the version of VS on the box. Right now we are close to shipping TFS 2015 Update 1. I think if we switch over to `ISetupConfiguration2`, we still need to fallback to `ISetupConfiguration` for older VS installs (otherwise break existing consumers).

Aside from a longer-term solution for the task, troubleshooting is my more immediate concern. Thank you for the pointers. I think I understand the required change for the above troubleshooting script now.

@Pitterling if repair doesn't work. I can produce a private copy of the task. Otherwise, switching to the MSBuild task and specifying the location of msbuild.exe will unblock. The main difference between the two tasks at this point is that VSBuild passes the `/p:VisualStudioVersion=15.0` (or whichever version)

**Comment by heaths** [comment]
@ericsciple you should be using the Microsoft.VisualStudio.Setup.Configuration.Interop NuGet package (embeddable types) instead of rolling your own. If you're using this in script form (i.e. what you're using above) you still need to declare the interfaces correctly. `ISetupConfiguration2` inherits from `ISetupConfiguration`, which means its VTBL contains all the methods of inherited interfaces as well as its own. I will email your internal address a pointer to our source and you can basically copy/paste that.

As for backcompat, you certainly can still use `ISetupConfiguration`, but the `IsLaunchable` that we use for VS is only available on `ISetupConfiguration2`, following COM rules of not changing shipped interfaces (though, I admit, with `ISetupConfiguration2` and other v2 interfaces I have been adding methods until we ship RTW because nothing is breaking since we don't support anyone implementing these interfaces (in that there's no supported scenarios) apart from unit tests).

**Comment by ericsciple** [comment]
@heaths Thanks! Yes I understand the ordering problem now (just affects troubleshooting script i pasted above, doesn't affect the logic used by the task).

I think updating to use the DLLs from NuGet is a good path forward. At this point we're late in the release cycle so I think after we branch for TFS 2017 Update 1. The current code used by the tasks pre-dates the NuGet package. At that time, it was either hand-roll or pull in a bunch of DLLs. I think at that time, I'll also need to investigate attempting to use `ISetupConfiguration2` and falling back to `ISetupConfiguration` for clients with older installs. It sounds like at the same time we should leverage `ISetupConfiguration2`'s `IsLaunchable`.

**Comment by heaths** [comment]
The newer query API DLLs actually use `IsLaunchable` by default when calling `ISetupConfiguration::EnumInstances`, so that may not be necessary if you're just enumerating launchable instances (which can have errors, but are otherwise installed with no reboots pending).

**Comment by heaths** [comment]
BTW, I published our official cmdlets that expose the query API this weekend. See https://github.com/Microsoft/vssetup.powershell. Basically, with PowerShellGet installed (default in WMF 5.0), run: `install-module vssetup -scope currentuser`

**Comment by heaths** [comment]
Installing Visual Studio 2017 or newer by design does not define a VSINSTALLDIR environment variable because of multi-instancing. So unless that file is processed by an application that already knows where it's associated instance is located and defines that environment variable in the process scope, that wouldn't work.
